/system/lt.h:48:29: error: cast between incompatible function types from

OS: Linux, gcc version 9.1.0
```
In file included from .../nothing/src/game/level/level_editor/player_layer.h:6,
                 from .../nothing/src/game/level/level_editor.c:10:
.../nothing/src/game/level/level_editor.c: In function 'level_editor_dump':
.../nothing/src/system/lt.h:48:29: error: cast between incompatible function types from 'int (*)(FILE *)' {aka 'int (*)(struct _IO_FILE *)'} to 'void (*)(void *)' [-Werror=cast-function-type]
   48 |     lt_push(lt, (void*)res, (Dtor)dtor)
      |                             ^
.../nothing/src/game/level/level_editor.c:268:22: note: in expansion of macro 'PUSH_LT'
  268 |     FILE *filedump = PUSH_LT(
      |                      ^~~~~~~
cc1: all warnings being treated as errors

```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,6 +46,8 @@ add_library(system STATIC
   src/dynarray.c
   src/hashset.h
   src/hashset.c
+  src/system/file.h
+  src/system/file.c
   )
 
 add_library(ebisp STATIC
@@ -163,8 +165,6 @@ add_executable(nothing
   src/game/level/level_editor/layer.c
   src/game/level/level_editor/label_layer.c
   src/game/level/level_editor/label_layer.h
-  src/system/file.h
-  src/system/file.c
 )
 target_link_libraries(nothing ${SDL2_LIBRARIES} system ebisp)
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -34,6 +34,8 @@ add_library(system STATIC
   src/system/log_script.c
   src/system/log_script.h
   src/system/lt.h
+  src/system/lt_adapters.h
+  src/system/lt_adapters.c
   src/system/nth_alloc.c
   src/system/nth_alloc.h
   src/system/stacktrace.c

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -179,7 +179,7 @@ add_executable(nothing_test
   )
 target_link_libraries(nothing_test ${SDL2_LIBRARIES} system ebisp)
 
-if(("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "CLANG"))
+if(("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU") OR ("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang"))
   set(CMAKE_C_FLAGS
     "${CMAKE_C_FLAGS} \
      -Wall \

--- a/default.nix
+++ b/default.nix
@@ -8,10 +8,6 @@ with import <nixpkgs> {}; {
                         pkgconfig
                         cmake
                         valgrind
-                        racket
-                        inotifyTools
-                        libxml2
-                        ninja
                       ];
     };
 }

--- a/default.nix
+++ b/default.nix
@@ -1,5 +1,5 @@
 with import <nixpkgs> {}; {
-    nothingEnv = stdenv.mkDerivation {
+    nothingEnv = gcc8Stdenv.mkDerivation {
         name = "nothing-env";
         buildInputs = [ stdenv
                         gcc

--- a/src/ebisp/parser.c
+++ b/src/ebisp/parser.c
@@ -10,13 +10,9 @@
 #include "ebisp/builtins.h"
 #include "ebisp/parser.h"
 #include "system/lt.h"
+#include "system/lt_adapters.h"
 
 #define MAX_BUFFER_LENGTH (5 * 1000 * 1000)
-
-static void fclose_lt(void* file)
-{
-    fclose(file);
-}
 
 static struct ParseResult parse_expr(Gc *gc, struct Token current_token);
 

--- a/src/game/level/boxes.c
+++ b/src/game/level/boxes.c
@@ -101,7 +101,6 @@ int boxes_update(Boxes *boxes,
                  float delta_time)
 {
     trace_assert(boxes);
-    trace_assert(delta_time);
 
     const size_t count = dynarray_count(boxes->body_ids);
     RigidBodyId *body_ids = dynarray_data(boxes->body_ids);

--- a/src/game/level/level_editor.c
+++ b/src/game/level/level_editor.c
@@ -478,7 +478,7 @@ static int level_editor_dump(const LevelEditor *level_editor)
     FILE *filedump = PUSH_LT(
         level_editor->lt,
         fopen(level_editor->file_name, "w"),
-        fclose);
+        fclose_lt);
 
     if (fprintf(filedump, "%s\n", level_metadata_title(level_editor->metadata)) < 0) {
         return -1;

--- a/src/game/level/level_editor.c
+++ b/src/game/level/level_editor.c
@@ -13,6 +13,7 @@
 #include "system/stacktrace.h"
 #include "system/nth_alloc.h"
 #include "system/lt.h"
+#include "system/lt_adapters.h"
 #include "system/log.h"
 #include "system/str.h"
 

--- a/src/game/level/level_editor/color_picker.c
+++ b/src/game/level/level_editor/color_picker.c
@@ -25,9 +25,11 @@ ColorPicker create_color_picker_from_rgba(Color color)
 {
     Color color_hsla = rgba_to_hsla(color);
     ColorPicker color_picker = {
-        .hue = {0, color_hsla.r, 360.0f},
-        .saturation = {0, color_hsla.g, 1.0f},
-        .lightness = {0, color_hsla.b, 1.0f}
+        .sliders = {
+            {0, color_hsla.r, 360.0f},
+            {0, color_hsla.g, 1.0f},
+            {0, color_hsla.b, 1.0f}
+        }
     };
     return color_picker;
 }
@@ -58,31 +60,15 @@ int color_picker_render(const ColorPicker *color_picker,
     trace_assert(camera);
 
     /* TODO(#931): Color Picker sliders don't have any labels */
-
-    if (slider_render(
-            &color_picker->hue,
-            camera,
-            rect(0.0f, COLOR_SLIDER_HEIGHT,
-                 COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT)) < 0) {
-        return -1;
+    for (ColorPickerSlider index = 0; index < COLOR_SLIDER_N; ++index) {
+        if (slider_render(
+                &color_picker->sliders[index],
+                camera,
+                rect(0.0f, COLOR_SLIDER_HEIGHT * (float) (index + 1),
+                     COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT)) < 0) {
+            return -1;
+        }
     }
-
-    if (slider_render(
-            &color_picker->saturation,
-            camera,
-            rect(0.0f, COLOR_SLIDER_HEIGHT * 2.0f,
-                 COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT)) < 0) {
-        return -1;
-    }
-
-    if (slider_render(
-            &color_picker->lightness,
-            camera,
-            rect(0.0f, COLOR_SLIDER_HEIGHT * 3.0f,
-                 COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT)) < 0) {
-        return -1;
-    }
-
 
     return 0;
 }
@@ -95,30 +81,15 @@ int color_picker_event(ColorPicker *color_picker, const SDL_Event *event, int *s
 
     int selected = 0;
 
-    if (slider_event(&color_picker->hue,
-                     event,
-                     rect(0.0f, COLOR_SLIDER_HEIGHT,
-                          COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT),
-                     &selected) < 0) {
-        return -1;
-    }
-
-    if (!selected) {
-        if (slider_event(&color_picker->saturation,
-                         event,
-                         rect(0.0f, COLOR_SLIDER_HEIGHT * 2.0f,
-                              COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT),
-                         &selected) < 0) {
-            return -1;
-        }
-    }
-
-    if (!selected) {
-        if (slider_event(&color_picker->lightness,
-                         event,
-                         rect(0.0f, COLOR_SLIDER_HEIGHT * 3.0f,
-                              COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT),
-                         &selected) < 0) {
+    for (ColorPickerSlider index = 0;
+         !selected && index < COLOR_SLIDER_N;
+         ++index) {
+        if (slider_event(
+                &color_picker->sliders[index],
+                event,
+                rect(0.0f, COLOR_SLIDER_HEIGHT * (float) (index + 1),
+                     COLOR_SLIDER_WIDTH, COLOR_SLIDER_HEIGHT),
+                &selected) < 0) {
             return -1;
         }
     }
@@ -133,8 +104,8 @@ int color_picker_event(ColorPicker *color_picker, const SDL_Event *event, int *s
 Color color_picker_rgba(const ColorPicker *color_picker)
 {
     return hsla(
-        color_picker->hue.value,
-        color_picker->saturation.value,
-        color_picker->lightness.value,
+        color_picker->sliders[COLOR_SLIDER_HUE].value,
+        color_picker->sliders[COLOR_SLIDER_SAT].value,
+        color_picker->sliders[COLOR_SLIDER_LIT].value,
         1.0f);
 }

--- a/src/game/level/level_editor/color_picker.h
+++ b/src/game/level/level_editor/color_picker.h
@@ -5,11 +5,16 @@
 #include "layer.h"
 #include "ui/slider.h"
 
+typedef enum {
+    COLOR_SLIDER_HUE = 0,
+    COLOR_SLIDER_SAT,
+    COLOR_SLIDER_LIT,
+    COLOR_SLIDER_N
+} ColorPickerSlider;
+
 typedef struct {
     // TODO(#933): ColorPicker should use array of sliders
-    Slider hue;
-    Slider saturation;
-    Slider lightness;
+    Slider sliders[COLOR_SLIDER_N];
 } ColorPicker;
 
 typedef struct LineStream LineStream;

--- a/src/game/level_folder.c
+++ b/src/game/level_folder.c
@@ -3,6 +3,7 @@
 
 #include "system/stacktrace.h"
 #include "system/lt.h"
+#include "system/lt_adapters.h"
 #include "system/nth_alloc.h"
 #include "system/line_stream.h"
 #include "system/str.h"
@@ -56,7 +57,7 @@ LevelFolder *create_level_folder(const char *dirpath)
     // TODO(#920): Level Picker is probably broken on Windows
     //   We have never tested our dirent implementation on Windows. So
     //   it's probably broken. We need to check it.
-    DIR *level_dir = PUSH_LT(lt, opendir(dirpath), closedir);
+    DIR *level_dir = PUSH_LT(lt, opendir(dirpath), closedir_lt);
 
     for (struct dirent *d = readdir(level_dir);
          d != NULL;

--- a/src/system/file.c
+++ b/src/system/file.c
@@ -13,6 +13,7 @@
 
 #include "system/stacktrace.h"
 #include "file.h"
+#include "lt_adapters.h"
 
 int last_modified(const char *filepath, time_t *time)
 {

--- a/src/system/line_stream.c
+++ b/src/system/line_stream.c
@@ -7,6 +7,7 @@
 
 #include "line_stream.h"
 #include "lt.h"
+#include "lt_adapters.h"
 #include "system/nth_alloc.h"
 #include "system/log.h"
 #include "system/str.h"
@@ -19,11 +20,6 @@ struct LineStream
     size_t capacity;
     bool unfinished;
 };
-
-static void fclose_lt(void* file)
-{
-    fclose(file);
-}
 
 // TODO(#905): create_line_stream probably does not need mode
 //   Because LineStream interface doesn't even have anything

--- a/src/system/lt_adapters.c
+++ b/src/system/lt_adapters.c
@@ -1,0 +1,14 @@
+#include <stdio.h>
+
+#include "file.h"
+#include "lt_adapters.h"
+
+void fclose_lt(void* file)
+{
+    fclose(file);
+}
+
+void closedir_lt(void *dir)
+{
+    closedir(dir);
+}

--- a/src/system/lt_adapters.h
+++ b/src/system/lt_adapters.h
@@ -1,0 +1,7 @@
+#ifndef LT_ADAPTERS_H_
+#define LT_ADAPTERS_H_
+
+void fclose_lt(void* file);
+void closedir_lt(void *dir);
+
+#endif  // LT_ADAPTERS_H_

